Add ruff to local tests and CI

jlapeyre opened the pull request:
This adds linting using `ruff` to the relevant configuration files. Only a few rules are enabled and none of them trigger an error in the current state of 	qiskit-terra.

The immediate goal of this PR is to test integrating `ruff` into local workflows as well as CI.  This is the reason why only very few rules are enabled, and none necessitated code changes.

If this PR is successful, more `ruff` rules will be added in subsequent PRs.

mtreinish commented on CONTRIBUTING.md:
```diff
@@ -422,25 +422,35 @@ Note: If you have run `test/ipynb/mpl_tester.ipynb` locally it is possible some
+running `tox -elint` which will run `black`, `ruff`, and `pylint` to check the
+local code formatting and lint. If black returns a code formatting error you can
+run `tox -eblack` to automatically update the code formatting to conform to the
+style. However, if `ruff` or `pylint` return any error you will have to fix
```
Ruff does have some auto fix rules which can automatically be fixed (to some degree at least) with `--fix`. But, probably not worth getting into that here.

Eric-Arellano replied:
No need to block this PR on this. But, I find it useful to have three top-level commands:

* `tox -e fmt`, run all formatters
* `tox -e fix`, run all formatters & fixers
* `tox -e lint`, run all formatters and fixers in check-only mode, along with pure linters

Why split `fix` and `fmt`? You sometimes only want to format your code without making any semantic changes. For example, some "fixers" will remove unused import. But that import might only be unused because you are still writing the code.

We've used `fmt` vs `fix` vs `lint` in Pantsbuild (a build system) for the past year and it's worked great.

jlapeyre replied:
And we have unused imports just for the side effects. I made a comment about fixing below. I agree you have to be a bit more careful with running fix. I think something like your three commands looks good. But maybe not for this PR.

mtreinish commented on CONTRIBUTING.md:
```diff
@@ -422,25 +422,35 @@ Note: If you have run `test/ipynb/mpl_tester.ipynb` locally it is possible some
+source tree, but makes up for this by being much faster (and those rare
+oversights will still be caught by the CI after you open a pull request).
+
+You can also run just `ruff`. If you have installed the development packages in
+your python environment via `pip install -r requirements.dev`, then `ruff` will
+be available. Otherwise install it with `pip install ruff`. Running the command
+`ruff check qiskit test tools examples setup.py` will run the same `ruff` tests
+that are run via `tox` as well as in CI.
```
I assume you put this here because ruff is fast and people might want to run it standalone. If so black fits into this category too.

jlapeyre replied:
Exactly. Some people (myself included) like to avoid `tox` when possible because it's a bit slow and complicated. Even though I run it at least once when everything is ready. And yes, adding a note about `black` makes sense and I don't think complicates the PR.

mtreinish reviewed (approved):
This LGTM, I think there are probably some other low hanging fruit rules we can add that are unlikely to be controversial or cause conflicts. But that's easy enough to do in a follow up. This starts us getting used to using the tool and we can incrementally ramp things up from here until we're ready to drop pylint. Just a couple questions inline but nothing blocking

Eric-Arellano reviewed (dismissed):
Thanks!

jlapeyre commented:
> Ruff does have some auto fix rules

I found they are not quite as fool proof as, say black. I don't recall the details. Maybe that we might want to make some exceptions. In any case, I tried to avoid complicating this first PR.

jlapeyre commented on requirements-dev.txt:
```diff
@@ -11,6 +11,7 @@ black[jupyter]~=22.0
 pydot
 astroid==2.14.2
 pylint==2.16.2
+ruff==0.0.267
```
Hmm dunno. I can check. I did see that `0.0.266` was withdrawn. But we can always deal with that kind of thing fairly easily.

mtreinish commented on CONTRIBUTING.md:
```diff
@@ -422,25 +422,36 @@ Note: If you have run `test/ipynb/mpl_tester.ipynb` locally it is possible some
+Because they are so fast, it is sometimes convenient to run the tools `black` and `ruff` separately
+rather than via `tox`. If you have installed the development packages in your python environment via
+`pip install -r requirements.dev`, then `ruff` and `black` will be available. Otherwise install them
+with `pip install ruff black`. Running the commands `ruff check qiskit test tools examples setup.py`
```
This is potentially problematic to suggest because of our pinning. If you run `pip install black` for example it will pull in a `23.x.y` version which will behave differently than what CI will check (I think we can look at bumping it in a separate PR as it's separate), but it might be good to include the pins here although I guess it's one more place to keep in sync.

Eric-Arellano replied:
I'd recommend we remove this paragraph for the exact problem you're describing.

One option is to have multiple environments:

- `tox -e fmt`
- `tox -e fix`
- `tox -e lint`
- `tox -e ruff`
- `tox -e black`
- `tox -e pylint`

The downside is that `tox.ini` will be more awkward, but that doesn't seem like a big deal to me. Optimize for developer experience, not how elegant a config file is that very few people touch.

Eric-Arellano replied:
> Won't tox -e ruff be much slower than just running ruff?

Why? It will be annoying the first time you have to install the virtual environment. But otherwise it should be roughly the same speed.

jlapeyre replied:
I'm looking for a good place in the contributors doc to warn users that they should install dev tools only via `requirements-dev.txt`. Or at least get the correct version if they do it piecemeal (btw, Having a robust packaging system instead of talking about this would be convenient.). I notice that `reno` is pinned, but there is an instruction to just `pip` install it.

jlapeyre replied:
I tried a `ruff` target in `tox.ini`. It runs in 900ms. Running `ruff` directly is 23ms. And less if you run it only on the directory you are editing. The story with `black` is similar... But don't think the comments I added there are optimal. I mean giving explicit instructions for one invocation of `ruff` at the cli doesn't really belong there.

Eric-Arellano commented on CONTRIBUTING.md:
```diff
@@ -448,10 +448,8 @@ oversights will still be caught by the CI after you open a pull request).
-with `pip install ruff black`. Running the commands `ruff check qiskit test tools examples setup.py`
-and `black --check qiskit test tools examples setup.py` will run the same tests with these tools
-that are run via `tox` as well as in CI.
+`pip install -r requirements-dev.txt`, then `ruff` and `black` will be available and can be run from
+the command line. See [`tox.ini`](tox.ini) for how `tox` invokes them.
```
Nice change. I like this.

jlapeyre commented:
* I simplified the discussion [`963958e` (#10116)](https://github.com/Qiskit/qiskit-terra/pull/10116/commits/963958e21eb23ea14ecffe0134b4386ecb6dbcc8)  of running `black` and `ruff` in CONTRIBUTING. I think a discussion of requirements-dev.txt would be a great idea, but that it would be a distraction in this PR.
* Comment on [reliability of ruff autofix](https://beta.ruff.rs/docs/faq/#ruff-tried-to-fix-something-but-it-broke-my-code)
* This is the only [discussion of organizing additions rules and default rules](https://github.com/charliermarsh/ruff/issues/2179) that I find. The bottom line seems the be that they are just adding stuff at 100 mph.
* At present running `ruff` remains hardcoded with no options into the existing `tox` test target. I also added a `Makefile` target. But I did not document it, again because whether and how to do that belongs in other issues and PRs. In general, I'm trying to dodge the question of how much API to offer to developers and TMTOWTDI.

Eric-Arellano reviewed (dismissed):
Thanks!

jlapeyre commented on pyproject.toml:
```diff
@@ -23,6 +23,14 @@ environment = 'RUSTUP_TOOLCHAIN="stable"'
 
+[tool.ruff]
+select = [
+       "F631",
+       "F632",
+       "F634",
+       "F823",
```
"But who will provide QA for the QA?"

Eric-Arellano replied:
Hah Prettier from JavaScript autoformats TOML and YAML files 😅

mtreinish reviewed (approved):
LGTM, thanks for doing this
